### ISSUE TITLE
Useless compile error on triple-equal 1|2 against 0

### ISSUE THREAD

**Issue by rogierschouten** [Microsoft/TypeScript#21724]
**TypeScript Version:**  2.7.1

**Search Terms:** infer, compare, equals

**Code**

```ts
type T = [number] | [number, number];  // now t.length has type 1|2
let t: T;
if (t!.length === 0) { // compile error: Operator '===' cannot be applied to types '1' and '0'
}
if (t!.length > 100) { // no compile error
}
if (t!.length === 1 - 1) { // no compile error
}
```

**Expected behavior:**
No compile errors on any of these. It's plain inconsistent and not useful.

**Actual behavior:**
Compile error on `t === 0` but not on other comparisons

**Playground Link:**
See above

**Related Issues:**

**Comment by andy-ms** [comment]
Tuples have a fixed length as of #17765. If you suspect that `t` could be a "tuple" of arbitrary length, you could simply use the type `number[]`, since tuples are arrays.
It's true that we could give an error on `t.length > 100` because all known lengths are less, but not sure if that would be worth the complexity to add. Same for evaluating constant expressions like `1 - 1` any operation on numbers is just treated as a `number` for simplicity.

**Comment by krryan** [comment]
For what it's worth, I would _love_ to see arithmetic constraints on things like lengths. But such an error only really becomes useful when users can write such constraints (I envision writing a function like `f(valueThatIsNotEmpty: string & { length: number > 0; })` or something).
